### ISSUE TITLE
pre-trained model

### ISSUE THREAD

**Issue by af-doom** [GuiyuZhao/SphereNet#1]
Hello, thank you very much for your great open source work. As a general descriptor, is it feasible to directly use the pre-trained model to migrate to my own dataset? What other parameters need to be adjusted besides the search radius?

**Comment by GuiyuZhao** [comment]
Thank you for your interest in our work. Yes, our method is known for its generalization, which can be applied to other datasets, which we have verified through a large number of experiments. For parameter Settings, you can adjust the radius size according to the object size of the data set to complete the generalization on other data sets. If there is a better requirement for accuracy, you can further fine-tune N,M,K or retrain the model.

**Comment by af-doom** [comment]
> 感谢您对我们的工作感兴趣。是的，我们的方法以其泛化性而闻名，这可以应用于其他数据集，我们已经通过大量实验验证了这一点。对于参数 Settings，您可以根据数据集的对象大小调整半径大小，以完成对其他数据集的泛化。如果对精度有更好的要求，您可以进一步微调 N，M，K 或重新训练模型。

Thank you for your quick reply. Hello, I tested the pre-trained model. I have a few questions: 1) Registration is basically impossible when there are a lot of outliers, because the same number of key points and descriptors are extracted for the source point cloud and the target point cloud（as shown in the figure below）. How should this be handled? Downsampling is ineffective. 2) Key point and descriptor extraction is very time-consuming, compared to traditional methods (such as FPFH+teaser++)
When there are a large number of outliers, multiple data tests are basically unsuccessful. The network does not seem to extract a reasonable number of descriptors based on the point cloud features (size, etc.), but rather as a fixed parameter. I wonder if this is the reason for the poor performance when there are a large number of outliers. Of course, your open source work is excellent enough, but I would like to ask you about this. For example:
![output](https://github.com/user-attachments/assets/025fd045-5b51-4fb1-b5b8-e5134e03f6e8)
![output (1)](https://github.com/user-attachments/assets/f6604dd2-ff96-45be-a387-8433bb71f79f)
A small number of outliers exist, such as two adjacent frames. Apart from the time consumption, the generalization performance of the network is excellent in the work I have tested. For example:
![image](https://github.com/user-attachments/assets/20596a7b-45d9-4148-8c47-d65f16afecf0)
![image](https://github.com/user-attachments/assets/22499dd6-07ef-4a59-b131-25bf3ef42807)

**Comment by af-doom** [comment]
> > Thank you for your interest in our work. Yes, our method is known for its generalization, which can be applied to other datasets, which we have verified through a large number of experiments. For parameter Settings, you can adjust the radius size according to the object size of the data set to complete the generalization on other data sets. If there is a better requirement for accuracy, you can further fine-tune N,M,K or retrain the model.
> > Thank you for your quick reply. Hello, I tested the pre-trained model. I have a few questions: 1) Registration is basically impossible when there are a lot of outliers, because the same number of key points and descriptors are extracted for the source point cloud and the target point cloud（as shown in the figure below）. How should this be handled? Downsampling is ineffective. 2) Key point and descriptor extraction is very time-consuming, compared to traditional methods (such as FPFH+teaser++) ![output](https://private-user-images.githubusercontent.com/91470350/395831832-025fd045-5b51-4fb1-b5b8-e5134e03f6e8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQyMzc1OTksIm5iZiI6MTczNDIzNzI5OSwicGF0aCI6Ii85MTQ3MDM1MC8zOTU4MzE4MzItMDI1ZmQwNDUtNWI1MS00ZmIxLWI1YjgtZTUxMzRlMDNmNmU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE1VDA0MzQ1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEwZWZlMDQzZmIxOTRlZDJkNjhmODljODQ4NDY1YWNmMzg5NDZlMDgyNjU5ZmQ1OTM1NWE5MDViMjk2ZWE1ZTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mG1nZ6ZaNy8gCXWlHFV26c2w8oYkgZEJJkrIStDdOqo)

**Comment by GuiyuZhao** [comment]
I think the reason why the examples you give fail is not the large number of outliers, but the low overlap and inconsistent point cloud field of view. That's my judgment based on your visualization, but I'm not sure. Can you give the visual result of ground-truth alignment?

If what I think is right, I think the solution is as follows:
The blue point cloud you give is large at the scene level, while the yellow point cloud is small at the object level. Therefore, we should not extract the same number of keys for both (the reason why the other examples you give succeed is because the two point clouds have roughly the same scale and number of points, so use the same number of keys). The solution is simple, extract a different number of key points for the source and target point clouds. For example, a yellow dot gives 1,000 keys, and a blue dot gives 5,000 keys (depending on the number of points in your point cloud). You need to modify the corresponding code.

**Comment by af-doom** [comment]
> I think the reason why the examples you give fail is not the large number of outliers, but the low overlap and inconsistent point cloud field of view. That's my judgment based on your visualization, but I'm not sure. Can you give the visual result of ground-truth alignment?
> 
> If what I think is right, I think the solution is as follows: The blue point cloud you give is large at the scene level, while the yellow point cloud is small at the object level. Therefore, we should not extract the same number of keys for both (the reason why the other examples you give succeed is because the two point clouds have roughly the same scale and number of points, so use the same number of keys). The solution is simple, extract a different number of key points for the source and target point clouds. For example, a yellow dot gives 1,000 keys, and a blue dot gives 5,000 keys (depending on the number of points in your point cloud). You need to modify the corresponding code.
Awesome, you are right. It is indeed very important to choose the number of keypoints according to the scene scale. Do you have any suggestions for adapting it?
![image](https://github.com/user-attachments/assets/901717e9-347b-4aa1-973b-ecb761618b5a)

**Comment by GuiyuZhao** [comment]
Based on your visualization, I was right. If the density of the two point clouds is roughly the same, you can set the ratio of key points according to the order of magnitude of the two point clouds.
In addition, because your target point cloud is muchg larger than source, if you can, you can do a pre-cut first for the target point cloud, which is similar to the work of place recognition.

**Comment by af-doom** [comment]
> Based on your visualization, I was right. If the density of the two point clouds is roughly the same, you can set the ratio of key points according to the order of magnitude of the two point clouds. In addition, because your target point cloud is muchg larger than source, if you can, you can do a pre-cut first for the target point cloud, which is similar to the work of place recognition. 

Thank you for your reply. I tested a few more groups. As a pre-trained model, the generalization performance is already very good. You are awesome. I have a question. How should I pre-cut without knowing any pose? Are there any corresponding papers to recommend? I have just started to study point cloud registration. For the registration of the visualized point cloud above,, the total time consumption was 70s, and the descriptor extraction took about 15s. Of course, the total time consumption also includes Ransac alignment and preprocessing, which is not within the scope of the paper. Do you have any suggestions for this?

**Comment by GuiyuZhao** [comment]
For the work of place recognition,  you can refer to https://openaccess.thecvf.com/content_ICCV_2019/papers/Liu_LPD-Net_3D_Point_Cloud_Learning_for_Large-Scale_Place_Recognition_and_ICCV_2019_paper.pdf. But I don't think you need to do the large-scene place recognition. 
Most of our time is spent on RANSAC. You can replace RANSAC with a more efficient oultier removal+SVD method. for details, you can refer to SC2-PCR: https://github.com/ZhiChen902/SC2-PCR

**Comment by af-doom** [comment]
> For the work of place recognition, you can refer to https://openaccess.thecvf.com/content_ICCV_2019/papers/Liu_LPD-Net_3D_Point_Cloud_Learning_for_Large-Scale_Place_Recognition_and_ICCV_2019_paper.pdf. But I don't think you need to do the large-scene place recognition. Most of our time is spent on RANSAC. You can replace RANSAC with a more efficient oultier removal+SVD method. for details, you can refer to SC2-PCR: https://github.com/ZhiChen902/SC2-PCR

Thank you very much for your reply.

**Comment by af-doom** [comment]
> For the work of place recognition, you can refer to https://openaccess.thecvf.com/content_ICCV_2019/papers/Liu_LPD-Net_3D_Point_Cloud_Learning_for_Large-Scale_Place_Recognition_and_ICCV_2019_paper.pdf. But I don't think you need to do the large-scene place recognition. Most of our time is spent on RANSAC. You can replace RANSAC with a more efficient oultier removal+SVD method. for details, you can refer to SC2-PCR: https://github.com/ZhiChen902/SC2-PCR

Your network is great. This is an global point cloud registration  implementation based on `SphereNet+teaser++.` Share it here and hope it will help spread your code. [SphereNet_TEASER](https://github.com/af-doom/SphereNet_TEASER)
